core/blockchain.go

```diff
@@ -1376,6 +1392,11 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 	rawdb.WriteBlock(blockBatch, block)
 	rawdb.WriteReceipts(blockBatch, block.Hash(), block.NumberU64(), receipts)
 	rawdb.WritePreimages(blockBatch, state.Preimages())
+	if (randomCommitment != common.Hash{}) {
```

Comment:
i'm hesitatnt to have this logic inside `blockchain`, would it make sense to have a blockchain listener that computes/writes the commitment for every block?

In fact, we could have a more cohesive type with all random related functions:

 * Listen to new Blocks to generate/write commitment
 * Reconstruct commitment cache
 * Get current / genearte current

Reply:
What's your hesitancy of putting this in `blockchain`?

The changes are following much of the same design that `blockchain` already implements.  E.g. the totalDifficulty is also a "cache", in that it's possible to reconstruct from the past chain, but it's computed at the time of block insert and saved in it's own leveldb location.

Reply:
+1 for keeping this in blockchain. I think keeping this code block here is much easier to understand than having a dedicated listener.

Reply:
My hesitation is about coupling...  or using other maxims: Separation of Concerns / Single Responsibility Principle

Now blockchain "knows" about randomness cache, (it's coupled to it). Same with miner.

And then, the "responsibility" of keeping the cache sane is split in a few places. There's the logic of the cache (read/write), there's the logic of check & recovery on miner, and there's the logic of generation with each new block on blockchain.

Now, in any future where a maintainer makes a change to those modules, it need to have the same logical context and awareness about this. And that's tricky, since it's just one of the many responsibilities these modules have.

That's why my suggestion is to have "something" that represents the randomness cache, and it's responsible of keeping it sane, recovering it and providing access to it. That "something" can even be an interface, thus it's easy to mock and test things in isolation.

@trianglesphere @kevjue what do you think?

Reply:
WDYT of keeping the randomness cache within the blockchain object (basically it would be a "primitive" of the block) and move the restoration logic from `miner` into `blockchain` (However, `miner` will still invoke the restoration logic if needed).

As for the actual insertion of the cache (at least for the normal case, as opposed to the restoration case), I still think that it should be done within the block insertion function (basically how it's implemented now), since it's inserted as part of a leveldb transaction along with all of the other parts of the block, since inserting all block related data atomically is one high level goal that we wanted to attain.

Reply:
I'm fine with bundling lines approx 1360 to 1370 into some manager object/primitive, but I agree with Kevin that `rawdb.WriteRandomCommitmentCache(...)` should stay here.

Reply:
I like it!